Preformatted block: Add support for font sizes

mendezcode opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/master/docs/contributors/repository-management.md#pull-requests. -->

## Description
Adds support for font sizes to the preformatted code block. It also changes the CSS for this block to use the global styles variable, setting a fallback font size.

## How has this been tested?
Tested with a block based and non block based version of `twentytwentyone`.

## Screenshots
![Screen Shot 2020-12-08 at 12 59 17 PM](https://user-images.githubusercontent.com/5881557/101515993-55d5a400-3955-11eb-82df-7e9f8ff81eb1.png)

## Types of changes
New feature (non-breaking change which adds functionality).

## Checklist:
- [ ] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

mendezcode committed:
Add preformatted block styles and font settings.

mendezcode committed:
Add font size fallback.

jasmussen commented on packages/block-library/src/preformatted/editor.scss:
```diff
@@ -0,0 +1,4 @@
+.wp-block-preformatted {
+	overflow-x: auto;
```
The `style.scss` file always gets loaded into the editor, so if a rule exists in that file, it does not need to be replicated in `editor.scss`.

mendezcode replied:
Oh, yes, true that. I'll fix accordingly.

jasmussen commented on packages/block-library/src/preformatted/editor.scss:
```diff
@@ -0,0 +1,4 @@
+.wp-block-preformatted {
+	overflow-x: auto;
+	white-space: pre !important;
```
Since this rule also exists in `style.scss`, we are so close to not needing any editor style at all for this block. And we should be avoiding !important anyway, if at all possible. I dug in to see why the !important was necessary, and it appears that `white-space: pre-wrap;` is applied as an inline style on the preformatted block:

<img width="619" alt="Screenshot 2020-12-09 at 08 20 30" src="https://user-images.githubusercontent.com/1204802/101597854-91c14580-39f7-11eb-8c28-a825bd58ec6e.png">

Why does that exist? I'm not sure, but from digging at the code, it looks to be related to this PR: #18656. Or possibly #17779. Specifically there's some commentary in https://github.com/WordPress/gutenberg/pull/17779/files#diff-7df6f5253ccec2f2e6d1bb22e81f87a4b6bccd2604463eeecded163aa27fd333R70 which seems relevant.

It seems we need to figure out why that output

mendezcode replied:
The `pre-wrap` inline style is coming from the `<Richtext>` component, and its value is defined here: https://github.com/WordPress/gutenberg/blob/master/packages/rich-text/src/component/index.js#L82

The value is set on the `whiteSpace` variable, which is then added to `defaultStyle`, which subsequently makes it to the component's inline styles via the `style` prop here: https://github.com/WordPress/gutenberg/blob/master/packages/rich-text/src/component/index.js#L1075

mendezcode replied:
According to the inspector, I can see that besides the `pre-wrap` inline style, there's the `pre-wrap` style attribute coming from the Block's `style.scss` stylesheet. If neither the inline style nor the block style are present, then, the `<pre>` element will get its default style from the browser defaults.

This makes me think we can live with the inline style, since the `<RichText>` component is using the style attribute to guarantee its proper display (my assumption). Assuming a future scenario where the `<RichText>` component removes the `pre-wrap` inline style, then we'll still be covered by the `pre-wrap` style available in the preformatted block's `style.scss`.

scruffian replied:
Why not just remove the inline style?

mendezcode replied:
That's a quite considerable change. The `<RichText>` component uses it by default, and it is used in many places. Changing this may change the overall behavior of the `<RichText>` component.

jasmussen commented:
Thanks so much for this PR! Here's the editor, before:

<img width="816" alt="editor before" src="https://user-images.githubusercontent.com/1204802/101598216-2035c700-39f8-11eb-936f-eaad76ab1210.png">

Frontend, before:

<img width="1028" alt="front before" src="https://user-images.githubusercontent.com/1204802/101598225-2330b780-39f8-11eb-8da2-9abc69b29f63.png">

Here's the editor, after:

<img width="847" alt="editor after" src="https://user-images.githubusercontent.com/1204802/101598237-26c43e80-39f8-11eb-8128-26bb9cc9fdb8.png">

Here's the frontend, after:

<img width="917" alt="front after" src="https://user-images.githubusercontent.com/1204802/101598250-2a57c580-39f8-11eb-81cd-4927ea2d0382.png">

The font size control is also added:

<img width="301" alt="Screenshot 2020-12-09 at 08 26 23" src="https://user-images.githubusercontent.com/1204802/101598314-4196b300-39f8-11eb-8ca9-8822c1f3c51b.png">

That font size stuff all seems fine. It also adds a `white-space: nowrap;` property with a horizontal scrollbar. While it's good that this makes the frontend and editor sync up, this is the part of the PR we need to tread carefully with:

1. It changes the default appearance of the block which is already in wide usage. While I generally think it's an okay change, it is nevertheless a change, which means we need to tread carefully.
2. As I also left some comments about, it changes the intrinsic `pre-wrap` property to `wrap`, which based on commentary in the code by @ellatrix might go against initial intentions.

It seems the safest way forward, which still puts the frontend in sync with the editor, is to go in the opposite direction.

Instead of:

```
.wp-block-preformatted {
	font-size: var(--wp--preset--font-size--extra-small, 1em);
	overflow-x: auto;
	white-space: pre;
}
```

you output

```
.wp-block-preformatted {
	font-size: var(--wp--preset--font-size--extra-small, 1em);
	white-space: pre-wrap;
}
```

That way you can also remove the editor style you added again, and we sync up the frontend. 

I'd also argue that this serves as a good bugfix for the lacklustre overflow handling.

What do you think?

mendezcode committed:
Remove unnecessary overflow property and use `pre-wrap` instead.

This is a more elegant solution compared to the previous one, which relied on overflow.

mendezcode commented:
> It changes the default appearance of the block which is already in wide usage. While I generally think it's an okay change, it is nevertheless a change, which means we need to tread carefully.

What I think would make sense here is to do the following:

Set the default font size to match the existing size in Gutenberg. Right now it is `--wp--preset--font-size--extra-small`, and it definitely looks different compared tho the size before. It may need to be set to the current value it has at the moment. I haven't looked, but there may be a bigger font size that matches the current size of the editor.

My thinking here is that this needs to be done as progressive enhancement instead of doing it like I did initially. It must have backwards compatibility with the existing preformatted block, which means it would be totally better for it to have the same size it had.

I'll work on the backwards compatibility fix for the font size and will keep you posted.

> What do you think?

Makes absolute sense, I pushed a commit to address the `pre-wrap` change. Thanks!

jasmussen commented:
> Set the default font size to match the existing size in Gutenberg. Right now it is --wp--preset--font-size--extra-small, and it definitely looks different compared to the size before. It may need to be set to the current value it has at the moment. I haven't looked, but there may be a bigger font size that matches the current size of the editor.

The size looks the same to me in my testing of TT1🔳s, but definitely worth testing the before and after in a few themes to be sure. 

> My thinking here is that this needs to be done as progressive enhancement instead of doing it like I did initially. It must have backwards compatibility with the existing preformatted block, which means it would be totally better for it to have the same size it had.

I think that's generally the safe mindset, but it doesn't have to be the case always. If the existing block has a bug or something in that category, it's fine to fix it. So the fact that it goes out to every user, is arguably a key benefit. So this isn't a fundamental issue, so much as us being careful that we do this deliberately and thinking through any potential side effects.

mendezcode committed:
Remove editor.scss

mendezcode commented:
I have removed editor.scss to leave only the `pre-wrap` coming from `style.scss`.
This does align with the `pre-wrap` present in the editor inline style.

nosolosw commented on packages/block-library/src/preformatted/style.scss:
```diff
@@ -0,0 +1,4 @@
+.wp-block-preformatted {
+	font-size: var(--wp--preset--font-size--extra-small, 1em);
```
:wave: We should remove the CSS Custom Property from here. I've seen that the code block recently got it added as well in https://github.com/WordPress/gutenberg/pull/27294/files#r540890153 and I've prepared a PR that fixes and explain why we shouldn't do that https://github.com/WordPress/gutenberg/pull/27672

nosolosw commented:
The changes to add the font-size look fine by me (other than removing the CSS Custom Property, see comment). I see there's some conversation about how to handle the wrapping, as soon as that's sorted out, it's fine to land this PR.

mendezcode committed:
Remove CSS custom property.

This is the theme's job. For more information, see #27672

jasmussen reviewed (approved):
This seems good to me! I'd love a sanity check from Andrés or Ben or Kjell, but otherwise, thank you for your work and congrats on your first PR!

scruffian reviewed (approved):
LGTM

scruffian commented:
We should also add this to the docs (docs/designers-developers/developers/themes/theme-json.md)

aristath commented on packages/block-library/src/preformatted/block.json:
```diff
@@ -12,6 +12,7 @@
-		"anchor": true
+		"anchor": true,
+		"fontSize": true
 	}
```
In https://github.com/WordPress/gutenberg/pull/25220 which was merged a few days ago, we allow block-styles to only be loaded when a block actually exists on a page for FSE themes. As a result of that change, when adding a stylesheet to a block we need to declare it in the `block.json` file.
Adding this line will allow FSE themes to load the stylesheet :+1: 
```suggestion
	},
	"style": "wp-block-preformatted"
```

mendezcode committed:
Add stylesheet to `block.json` file

This is added due to new requirement implemented in #25220

Co-authored-by: Ari Stathopoulos <aristath@gmail.com>

mendezcode commented:
> We should also add this to the docs (docs/designers-developers/developers/themes/theme-json.md)

True that. I'll create a PR for it after this one is merged.

mendezcode committed:
Add preformatted block to theme-json.md specs.

mendezcode commented on docs/designers-developers/developers/themes/theme-json.md:
```diff
@@ -330,6 +330,7 @@ These are the current color properties supported by blocks:
 | Post Hierarchical Terms | Yes | Yes | Yes | Yes |
 | Post Tags | Yes | Yes | Yes | Yes |
 | Post Title | Yes | Yes | - | Yes |
```
This is actually wrong, I'm changing this in the next commit.

mendezcode committed:
Fix preformatted block support information on theme-json.md specs.
